# Exploratory Data Analysis
## Data Overview
**Goal**: Understand the structure of the dataset, including the number of rows, columns, and data types.

**Key Steps**: Load the data and print a concise summary of the dataset using .info() and .head().

In [1]:
#Import necessary libraries
import pandas as pd
import logging
import os, sys
# Add the 'scripts' directory to the Python path for module imports
sys.path.append(os.path.abspath(os.path.join('..', 'scripts')))
# Import load_data module
from load_data import load_data # type: ignore

# Set max rows and columns to display
pd.set_option('display.max_columns', 200)
pd.set_option('display.max_rows', 200)

# Configure logging
from custom_logger import setup_logger # type: ignore
logger = setup_logger()

logger.info("Imported necessary libraries.")

2025-01-23 15:57:22,294 - INFO - Imported necessary libraries.


Load the dataset

In [2]:
logger.info("Data loading initiated.")
df = load_data('../data/data.csv')  # Assume load_data() is your function
logger.info("Data loaded successfully.")

2025-01-23 15:57:25,132 - INFO - Data loading initiated.
2025-01-23 15:57:25,414 - INFO - Data loaded successfully.


Data successfully loaded from ../data/data.csv
Dataset contains 95662 rows and 15 columns.



### 1. Data overview:
Provide an overview of the dataset including shape, data types, missing values, and first few rows.

In [9]:
# Import the class CreditRiskEDA
from credit_eda_analysis import CreditRiskEDA
# Initialize the class
cr_eda = CreditRiskEDA(df)
# Logging activity
logger.info("Data overview initiated.")

# Overview of the dataset
if not df.empty:
    cr_eda.data_overview()

logger.info("Data overview successfully completed.")

2025-01-23 16:01:53,052 - INFO - Data overview initiated.


Data Overview:
Number of rows: 95662
Number of columns: 15

Column Data Types:
BatchId                              object
AccountId                            object
SubscriptionId                       object
CustomerId                           object
CurrencyCode                         object
CountryCode                           int64
ProviderId                           object
ProductId                            object
ProductCategory                      object
ChannelId                            object
Amount                              float64
Value                                 int64
TransactionStartTime    datetime64[ns, UTC]
PricingStrategy                       int64
FraudResult                           int64
dtype: object

First Five Rows:


BatchId       AccountId       SubscriptionId  \
TransactionId                                                              
TransactionId_76871   BatchId_36123  AccountId_3957   SubscriptionId_887   
TransactionId_73770   BatchId_15642  AccountId_4841  SubscriptionId_3829   
TransactionId_26203   BatchId_53941  AccountId_4229   SubscriptionId_222   
TransactionId_380    BatchId_102363   AccountId_648  SubscriptionId_2185   
TransactionId_28195   BatchId_38780  AccountId_4841  SubscriptionId_3829   

                          CustomerId CurrencyCode  CountryCode    ProviderId  \
TransactionId                                                                  
TransactionId_76871  CustomerId_4406          UGX          256  ProviderId_6   
TransactionId_73770  CustomerId_4406          UGX          256  ProviderId_4   
TransactionId_26203  CustomerId_4683          UGX          256  ProviderId_6   
TransactionId_380     CustomerId_988          UGX          256  ProviderId_1   
TransactionId_28195   CustomerId_988          UGX          256  ProviderId_4   

                        ProductId     ProductCategory    ChannelId   Amount  \
TransactionId                                                                 
TransactionId_76871  ProductId_10             airtime  ChannelId_3   1000.0   
TransactionId_73770   ProductId_6  financial_services  ChannelId_2    -20.0   
TransactionId_26203   ProductId_1             airtime  ChannelId_3    500.0   
TransactionId_380    ProductId_21        utility_bill  ChannelId_3  20000.0   
TransactionId_28195   ProductId_6  financial_services  ChannelId_2   -644.0   

                     Value      TransactionStartTime  PricingStrategy  \
TransactionId                                                           
TransactionId_76871   1000 2018-11-15 02:18:49+00:00                2   
TransactionId_73770     20 2018-11-15 02:19:08+00:00                2   
TransactionId_26203    500 2018-11-15 02:44:21+00:00                2   
TransactionId_380    21800 2018-11-15 03:32:55+00:00                2   
TransactionId_28195    644 2018-11-15 03:34:21+00:00                2   

                     FraudResult  
TransactionId                     
TransactionId_76871            0  
TransactionId_73770            0  
TransactionId_26203            0  
TransactionId_380              0  
TransactionId_28195            0

2025-01-23 16:01:53,096 - INFO - Data overview successfully completed.



Missing Values Overview:
BatchId                 0
AccountId               0
SubscriptionId          0
CustomerId              0
CurrencyCode            0
CountryCode             0
ProviderId              0
ProductId               0
ProductCategory         0
ChannelId               0
Amount                  0
Value                   0
TransactionStartTime    0
PricingStrategy         0
FraudResult             0
dtype: int64


### Data Overview Summary

+ The dataset contains 95,662 rows and 15 columns, providing a substantial amount of data for analysis.

+ The data types are generally correct, except for the TransactionStartTime column, which is currently stored as an object. This column should be converted to datetime format for accurate time-based analysis.

+ No missing values were detected in any of the columns, ensuring the dataset is complete and ready for further analysis without the need for imputation.

In [10]:
# Convert the TransactionStartTime to appropriate datetime format
df['TransactionStartTime'] = pd.to_datetime(df['TransactionStartTime'])

### 2. Summary Statistics

Understand the central tendency, dispersion, and shape of the dataset’s distribution.

In [11]:
# Log the start of the summary statistics process
logger.info("Generating summary statistics for numeric columns...")   

# Statistical summary
summary_stats = cr_eda.summary_statistics()
display(summary_stats.T)
# Log completion
logger.info("Summary statistics generation completed.")

2025-01-23 16:02:01,302 - INFO - Generating summary statistics for numeric columns...


CountryCode        Amount         Value  PricingStrategy  \
count         95662.0  9.566200e+04  9.566200e+04     95662.000000   
mean            256.0  6.717846e+03  9.900584e+03         2.255974   
std               0.0  1.233068e+05  1.231221e+05         0.732924   
min             256.0 -1.000000e+06  2.000000e+00         0.000000   
25%             256.0 -5.000000e+01  2.750000e+02         2.000000   
50%             256.0  1.000000e+03  1.000000e+03         2.000000   
75%             256.0  2.800000e+03  5.000000e+03         2.000000   
max             256.0  9.880000e+06  9.880000e+06         4.000000   
median          256.0  1.000000e+03  1.000000e+03         2.000000   
mode            256.0  1.000000e+03  1.000000e+03         2.000000   
skewness          0.0  5.109849e+01  5.129109e+01         1.659057   
kurtosis          0.0  3.363128e+03  3.378069e+03         2.087534   

           FraudResult  
count     95662.000000  
mean          0.002018  
std           0.044872  
min           0.000000  
25%           0.000000  
50%           0.000000  
75%           0.000000  
max           1.000000  
median        0.000000  
mode          0.000000  
skewness     22.196295  
kurtosis    490.685762

2025-01-23 16:02:01,360 - INFO - Summary statistics generation completed.


### Observations:

+ The CountryCode is constant at 256, indicating no variability in this column.

+ The Amount and Value features show a high degree of skewness (both over 51), indicating potential outliers or heavy-tailed distributions.

+ PricingStrategy is fairly stable with most values centered around 2.

+ FraudResult has a very low mean, indicating a low occurrence of fraudulent transactions, with a highly positive skewness suggesting a majority of zeros.

### Distribution of Numerical Features

+ Visualize the distribution of numerical features to identify patterns, skewness, and potential outliers.

In [6]:
# Import the class CreditRiskEDA
from credit_eda_visualize import CreditRiskEDAVisualize
# Initialize the class
cr_eda_visual = CreditRiskEDAVisualize(df)
# Logger activity
logger.info("Plot Numberical Distribution...")
# List of numeric columns
numeric_cols = df.select_dtypes(include='number').columns

# Plot distibution
cr_eda_visual.plot_numerical_distribution(numeric_cols)
logger.info("The distribution plot successfully completed.")

ModuleNotFoundError: No module named 'matplotlib'